### DISCUSSION TITLE
Grain boundaries are not reconstructed when using `calcGrains`

### DISCUSSION THREAD

**Discussion by PengWei97** [mtex-toolbox/mtex#1509] [Ask Anything] [answered]
Dear MTEX experts,

Recently, I want to count the grain size data of pure titanium based on the data results of ebsd, but I encountered some problems when carrying out grain reconstruction. Specifically, it is the esbd obtained by executing the command `grains = calcGrains(ebsd('Ti-Hex'),'theshold',0.1*degree)` and then executing `plot(grains.boundary,'lineWidth',2)`. The grains at region B in the figure are not reconstructed well, even though I set theshold equal to 0.01*degree.
![image](https://user-images.githubusercontent.com/56183775/181680355-f94a663f-9550-47a5-a1ec-a834d0160610.png)

So as a mtex beginner, I don't know how to solve this problem. Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by kilir** [comment]
Hi Wei,
what happens if you do `grains = calcGrains(ebsd)`?
Assuming you use a current version of Mtex, what does `grains = calcGrains(ebsd,'Qhull')` do?
Also, can you post the code you used for the two different grain maps (both of which do not separate a/b)?
Cheers,
Rüdiger

**Reply by PengWei97** [reply]
Oh, Thanks for your reply.
According to the suggestion, I execute the following code, and the result is,
```matlab
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');
ebsd = ebsd(inpolygon(ebsd,[80 420 120 100]));
grains = calcGrains(ebsd);

plot(grains.boundary,'lineWidth',2)
```

Then execute the second calcGrains configuration,
```matlab
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');
ebsd = ebsd(inpolygon(ebsd,[80 420 120 100]));
grains = calcGrains(ebsd,'Qhull');

plot(grains.boundary,'lineWidth',2)
```
![image](https://user-images.githubusercontent.com/56183775/181730303-c72bcaeb-7294-4042-8b76-f2396ff44975.png)

By comparing the above two pictures, it is found that there is no obvious difference. By the way, the version I am using is mtex-5.8.1.

> Also, can you post the code you used for the two different grain maps (both of which do not separate a/b)?

Here is the complete code from the previous conversation I run,

```matlab
clear all
close all

% crystal symmetry
CS = {... 
  'notIndexed',...
  crystalSymmetry('6/mmm', [3 3 4.7], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Ti-Hex', 'color', [0.53 0.81 0.98])};

% plotting convention
setMTEXpref('xAxisDirection','north');
setMTEXpref('zAxisDirection','outOfPlane');

%% Specify File Names

% path to files
pname = 'F:\simulationResult\pengweiGNG2022\experiment\lixiao\EBSD源文件\TI原位\10min';

% which files to be imported
fname = [pname '\Tituihuo 700du10min.crc'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');
ebsd = ebsd(inpolygon(ebsd,[80 420 120 100]));
plotx2east

figure(1)
% plot(ebsd('Ti-Hex'),ebsd('Ti-Hex').orientations,'coordinates','on')
plot(ebsd,ebsd.orientations,'coordinates','on')

grains = calcGrains(ebsd('Ti-Hex'),'theshold',0.1*degree);
ebsd(grains(grains.grainSize<20)) = [];
grains = smooth(grains,50);
hold on
plot(grains.boundary,'lineWidth',2)
hold off

figure(2)
[grains,ebsd.grainId] = calcGrains(ebsd('Ti-Hex'),'threshold',[1*degree, 15*degree]);
plot(grains,'lineWidth',2)
```

wei

**Reply by PengWei97** [reply]
In addition, I learned that the interface not identified by region A is the sub-grain boundary or low angle GB. I can identify the sub-grain boundary by setting the sub-grain boundary. The code and result are as follows,
```matlab
figure(3)
[grains,ebsd.grainId] = calcGrains(ebsd('Ti-Hex'),'threshold',[1*degree, 15*degree]);

grains = smooth(grains,50)
plot(ebsd,ebsd.orientations)
hold on
plot(grains.boundary,'linewidth',2)
alpha = grains.innerBoundary.misorientation.angle / (5*degree);
plot(grains.innerBoundary,'linewidth',1.5,'edgeAlpha',alpha,'linecolor','r');
hold off
```

![image](https://user-images.githubusercontent.com/56183775/181734009-8fea8c70-2c3d-4a36-bd23-88dae7b84d4b.png)

Although the sub-grain boundary in region A can be identified, this region A is still a grain. How to divide this large region into multiple small grains according to the grain boundary and sub-grain boundary?

**Reply by kilir** [reply]
Actually the inner boundaries look pretty much as if they'd  connect the outer boundaries. Hence a sufficiently small threshold > 1degree should still  pick them up as grain boundaries.

How does 
```
[grains,ebsd.grainId] = calcGrains(ebsd('Ti-Hex'),'threshold',[1*degree, 15*degree]);

gbt = grains.boundary('T','T');
plot(gbt,gbt.misorientation.angle/degree,'linewidth',2)
hold on
alpha = grains.innerBoundary.misorientation.angle/degree;
plot(grains.innerBoundary,alpha);
hold off
mtexColorbar
```
look like?

**Reply by PengWei97** [reply]
> Hence a sufficiently small threshold > 1degree should still pick them up as grain boundaries.

Yeah, if the threshold angle is set to 0.01*degree, the result still does not recognize these low angle grain boundaries as grain boundaries.

> How does `...` look like?

According to the code you provided, the output is,
![image](https://user-images.githubusercontent.com/56183775/181865330-deb59563-6284-4e5c-9530-6b9ae66b95e2.png)

**Reply by kilir** [reply]
Could you share this piece of dataset?

**Reply by PengWei97** [reply]
> Could you share this piece of dataset?

This is the original ebsd file of the area shown above,

[local_alpha-Ti.zip](https://github.com/mtex-toolbox/mtex/files/9227905/local_alpha-Ti.zip)

**Comment by kilir** [comment, marked as answer]
I do not seem to be able to reproduce your problem.
![ex](https://user-images.githubusercontent.com/11697557/182018299-92cc9562-8c29-43a3-a951-c3a1498bd45b.png)

I think using `'threshold'`  instead of `'theshold'` might help.



Cheers,
Rüdiger

**Reply by PengWei97** [reply]
Oh! You are right. I probably didn't write the `threshold` correctly when I first set it up, but I modified it later, and the result is very consistent with yours. 

Thank you very much for your enthusiastic answer, please forgive me for making such a stupid mistake.

wei
